### PR TITLE
apiserver: avoid global provider and image source registry

### PR DESCRIPTION
**Pull request by rogpeppe**
apiserver: avoid global provider and image source registry

This means we can pass in the registries when creating
an API server and so potentially have two different API
servers at once both using different providers/image sources.
It also means we can easily create a provider for
a specific test with specific properties rather than relying
on provider/dummy as a one-size-fits-all test provider.

The tests still use the global registries, to avoid excess churn.

I ensured that there were no remaining dependencies on the global
registries by compiling apiserver/... with the relevant functions from
environs removed.

I've split this PR into two commits - one that adds the necessary
registry type to environs, and the other (lots of obvious changes)
that makes apiserver use it.

### CODE REVIEW

**Comment by rogpeppe** [comment]
The test failure is an unrelated sporadic failure that also happens on tip: https://bugs.launchpad.net/juju/+bug/1744579

**Comment by axw** [comment]
I have had a cursory review (which is all I can currently afford), and this looks good in principle. Big +1 to injecting registries, and not relying on globals. Would be great to remove them altogether in another PR updating the tests.

File: apiserver/facades/agent/upgrader/upgrader_test.go [hunk 1]
```diff
@@ -102,7 +110,13 @@ func (s *upgraderSuite) TestWatchAPIVersion(c *gc.C) {
 func (s *upgraderSuite) TestUpgraderAPIRefusesNonMachineAgent(c *gc.C) {
 	anAuthorizer := s.authorizer
 	anAuthorizer.Tag = names.NewUnitTag("ubuntu/1")
-	anUpgrader, err := upgrader.NewUpgraderAPI(s.State, s.resources, anAuthorizer)
+	anUpgrader, err := upgrader.NewUpgraderAPI(facadetest.Context{
```
Comment (jameinel) [inline apiserver/facades/agent/upgrader/upgrader_test.go]: given the only thing changing between the tests is the authorizer, would it be worth pulling that out into a helper?
Not necessary, just odd to see the same thing over and over again. Its quite a chunk of text.

File: controller/config.go [hunk 2]
```diff
@@ -607,43 +607,43 @@ func GenerateControllerCertAndKey(caCert, caKey string, hostAddresses []string)
 }
 
 var configChecker = schema.FieldMap(schema.Fields{
+	AllowModelAccessKey:     schema.Bool(),
+	APIPort:                 schema.ForceInt(),
```
Comment (jameinel) [inline controller/config.go]: I assume this is just a Sort() on the keys?

**Comment by mitechie** [comment]
Hello, I am the PR robot. Your PR is older
    than 14 days. I'm closing it for now. If
    it needs attention please work with folks
    to progress the PR and land it.
    Beep..Boop...

**Pull request closed without merging**